MagicalCryptoWallet/Services/IndexBuilderService.cs

```diff
@@ -179,134 +187,160 @@ public void Syncronize()
-							.Select(entry => entry.Key.Hash + ":" + entry.Key.N + ":" + ByteHelpers.ToHex(entry.Value.ToCompressedBytes())));
+			if (prevHash != null)
+			{
+				// In case of reorg:
+				if (prevHash != block.Header.HashPrevBlock)
+				{
+					Logger.LogInfo<IndexBuilderService>($"REORG Invalid Block: {prevHash}");
+					// 1. Rollback index
+					using (await IndexLock.LockAsync())
+					{
+						Index.RemoveAt(Index.Count - 1);
+					}
+
+					// 2. Serialize Index. (Remove last line.)
+					await indexFile.FlushAsync();
+					indexFile.SetLength(_indexPositions.Pop());
 
-						Logger.LogInfo<IndexBuilderService>($"Created filter for block: {height}.");
+					// 3. Rollback Bech32UtxoSet
+					Bech32UtxoSetHistory.Rollback(Bech32UtxoSet); // The Bech32UtxoSet MUST be recovered to its previous state.
+
+					// 4. Serialize Bech32UtxoSet.
+					await SaveBech32UtxoAsync();
+
+					// 5. Skip the current block.
+					return;
+				}
+			}
+
+			Bech32UtxoSetHistory.ClearActionHistory(); //reset history.
+
+			var scripts = new HashSet<Script>();
+
+			foreach (var tx in block.Transactions)
+			{
+				for (int i = 0; i < tx.Outputs.Count; i++)
+				{
+					var output = tx.Outputs[i];
+					if (!output.ScriptPubKey.IsPayToScriptHash && output.ScriptPubKey.IsWitness)
+					{
+						var outpoint = new OutPoint(tx.GetHash(), i);
+						Bech32UtxoSet.Add(outpoint, output.ScriptPubKey);
+						Bech32UtxoSetHistory.StoreAction(ActionHistoryHelper.Operation.Add, outpoint, output.ScriptPubKey);
+						scripts.Add(output.ScriptPubKey);
 					}
-					catch (Exception ex)
+				}
+
+				foreach (var input in tx.Inputs)
+				{
+					var found = Bech32UtxoSet.ContainsKey(input.PrevOut);
+					if (found)
 					{
-						Logger.LogDebug<IndexBuilderService>(ex);
+						Script val = Bech32UtxoSet[input.PrevOut];
+						Bech32UtxoSet.Remove(input.PrevOut);
+						Bech32UtxoSetHistory.StoreAction(ActionHistoryHelper.Operation.Remove, input.PrevOut, val);
+						scripts.Add(val);
 					}
 				}
-			});
+			}
+
+			// https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki
+			// The parameter k MUST be set to the first 16 bytes of the hash of the block for which the filter 
+			// is constructed.This ensures the key is deterministic while still varying from block to block.
+			var key = block.GetHash().ToBytes().SafeSubarray(0, 16);
+
+			GolombRiceFilter filter = null;
+			if (scripts.Count != 0)
+			{
+				filter = GolombRiceFilter.Build(key, scripts.Select(x => x.ToBytes()));
+			}
+
+			var filterModel = new FilterModel
+			{
+				BlockHash = block.GetHash(),
+				BlockHeight = new Height(height),
+				Filter = filter
+			};
+
+			_indexPositions.Push(indexFile.Position);
+			var buff = Encoding.ASCII.GetBytes(filterModel.ToLine() + Environment.NewLine);
+			await indexFile.WriteAsync(buff, 0, buff.Length);
+
+			using (await IndexLock.LockAsync())
+			{
+				Index.Add(filterModel);
+			}
+
+			if(height % 10000 == 0)
```

Comment:
I don't think it's good. It must be always in sync with the index file.  

Imagine this: next time you load the both files the same time, the index will be more advanced than the utxoset.

Reply:
you're right. I will go back to this.